### ISSUE TITLE
Llama3.2 3B accuracy is lower with MHA2MLA

### ISSUE THREAD

**Issue by KMS07** [JT-Ushio/MHA2MLA#24]
I have fine-tuned the Llama3.2 3B model with MHA2MLA (r = 4, dkv = 8). The results of the CS ability show a very high-accuracy drop.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/mkotthag/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/mkotthag/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/mkotthag/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-IN style='tab-interval:36.0pt;word-wrap:break-word'>

Model | hellaswag | arc | piqa | winogrande | openbook_qa | mmlu
-- | -- | -- | -- | -- | -- | --
Llama3.2-3BMHA2MLA | 49.65 | 45.45 | 71 | 52.4 | 35.2 | 31.54
Llama3.2-3Boriginal | 70.75 | 59.1 | 77.85 | 64.95 | 41.6 | 42

</body>

</html>

Data used is the Smollm-corpus. I think this is the cause of it.

**Comment by KMS07** [comment]
Could you please look into this @JT-Ushio

**Comment by JT-Ushio** [comment]
Sorry for the late reply.

The gap between the MHA2MLA fine-tuning data and the original data may lead to potential performance loss. Regarding the significant performance degradation observed on Llama3.2, I have two suggestions:

1. Please take a look at the results of the original Llama3.2 3B model after continued training directly on Smollm-corpus. If a similar performance loss is observed, this would indicate a significant data gap between `Llama3.2` and `Smollm-corpus`.

2. If (1) holds true, then we should improve the quality of the MHA2MLA fine-tuning data (ideally, it would be best if Llama3.2 could provide a subset of their training data😥).

**Comment by JT-Ushio** [comment]
@KMS07 Thanks for the info!

@bg51717 — could you please run MHA2MLA on Llama3.2‑3B and check whether the **default partial‑rope ratio** and/or the **low‑rank (dkv) settings** are causing the large accuracy drop? If they are, could you try increasing the partial‑rope ratio and/or the low‑rank dimensions (e.g., larger `r` or `dkv`) and report the results?

**Comment by KMS07** [comment]
Hi @bg51717, sorry for the delayed response... I was busy.

This is the config for continued training:

```
# TrainingArguments
seed: 42
max_steps: 12000
report_to: wandb
run_name: llama3-2-3b_org_a6000-4GPU
save_strategy: steps
save_steps: 0.1
output_dir: ckpts/llama3-2-3b_org_a6000-4GPU
overwrite_output_dir: true
logging_strategy: steps
logging_steps: 20
resume_from_checkpoint: null
per_device_train_batch_size: 10
remove_unused_columns: False
gradient_accumulation_steps: 4
fp16: true
deepspeed: cfgs/ds_zero_2.json
dataloader_drop_last: true
# new args
# warmup_steps: 100
dataloader_pin_memory: true
dataloader_num_workers: 64
gradient_checkpointing: true

# optim
optim: adamw_torch
# lr_warmup_steps: 1200     
max_grad_norm: 1.0
learning_rate: 1.0e-5
adam_beta1: 0.9
adam_beta2: 0.95
adam_epsilon: 1.0e-8
weight_decay: 0.0
# is_freeze_non_attn: true
# lr scheduler
use_constant_with_warmup_decay_scheduler: true
lr_scheduler_kwargs:
  {
    "lr_decay_starting_step": 10000,
    "lr_decay_steps": 2000,
    "lr_decay_style": "1-sqrt",
    "lr_warmup_steps": 1200,
    "lr_warmup_style": "linear",
    "min_decay_lr": 0,
  }

# ModelArguments
model_name_or_path: meta-llama/Llama-3.2-3B
partial_rope_version: 2-norm # 'high'/'low'/'uniform'/'2-norm'
rope_dim_for_mla: 8
uniform_start_point: 0 # optional only for 'uniform'
# num_query_heads_div_key_heads: 1
qk_tensor_path: utils/llama3.2-3B_2norm_rank.pth
svd_init_method: joint # 'split'/'joint'
low_rank: 8
is_baseline: false
is_gqa2mha2mla: false

# DataArguments
is_nanoset: true
dataset_folders:
  - /home/kms/MHA2MLA/data/llama3_2_3B/fineweb-edu-dedup/
  - /home/kms/MHA2MLA/data/llama3_2_3B/cosmopedia-v2/
  - /home/kms/MHA2MLA/data/llama3_2_3B/python-edu/
  -/home/kms/MHA2MLA/data/llama3_2_3B/open-web-math/
  - /home/kms/MHA2MLA/data/llama3_2_3B/stackoverflow/
dataset_weights:
        - 0.7
        - 0.15
        - 0.08
        - 0.06
        - 0.01
sequence_length: 2048
```
